astropy/coordinates/builtin_frames/utils.py

```diff
@@ -236,25 +279,61 @@ def atciqz(rc, dc, astrom):
     same approach used by the ERFA functions above, except that they use a threshold of
     9 arcseconds.
 
+    In addition, ERFA's atciqz assumes a distant source, so there is no difference between
+    the object-Sun vector and the observer-Sun vector. This can lead to errors of up to a
+    few arcseconds in the worst case (e.g a Venus transit).
+
     Parameters
     ----------
-    rc : float or `~numpy.ndarray`
-        right ascension, radians
-    dc : float or `~numpy.ndarray`
-        declination, radians
+    astrometric : `~astropy.coordinates.CartesianRepresentation`
+        Astrometric ICRS position of object from observer
     astrom : eraASTROM array
         ERFA astrometry context, as produced by, e.g. ``eraApci13`` or ``eraApcs13``
 
     Returns
     --------
     ri : float or `~numpy.ndarray`
+        Right Ascension in radians
     di : float or `~numpy.ndarray`
+        Declination in radians
     """
+    # first of all get distance from object to observer
+    object_observer_distance = astrometric.norm()
+    # ignore parallax effects if no distance, or far away
+    ignore_distance = object_observer_distance.unit == u.one
+
     # BCRS coordinate direction (unit vector).
-    pco = erfa.s2c(rc, dc)
+    if not ignore_distance and u.allclose(astrometric.xyz, 0*u.km):
+        # special casing for positions at origin of frame
+        # replicating the behaviour of erfa.s2c with 0, 0 input
+        # which for reasons unknown gives the vector (1, 0, 0)
+        xhat = astrometric.unit_vectors()['x']
+        tmp = astrometric + 1*astrometric.x.unit*xhat
+        pco = (tmp/tmp.norm()).get_xyz(xyz_axis=-1).value
+    else:
+        pco = (astrometric / object_observer_distance).get_xyz(xyz_axis=-1).value
+
+    # Find BCRS direction of Sun to object
+    if ignore_distance:
+        # No distance to object, assume a long way away
+        q = pco
+    else:
+        # astrom['eh'] and astrom['em'] contain Sun to observer unit vector,
+        # and distance, respectively.
+        eh = astrom['em'] * CartesianRepresentation(astrom['eh'], unit=u.au, xyz_axis=-1, copy=False)
+        # apply parallax to find unit vector from Sun to object
+        q = eh + astrometric
```

Comment:
Sorry to be so late here. I just noticed that you had already made an attempt at this. The problem here is that `eh` is the non-light delayed position of the Earth wrt the Sun. The Earth moves fairly quickly in its orbit, so it isn't valid to combine the the light delayed position of the target body wrt to the Earth and the non-light time delayed position of the Earth wrt to the Sun. For the proper way to do this please see my (late) comments at #10291.

Reply:
To do this properly from a topocentric perspective, you'd need the position of the observer wrt to the Sun time delayed by the time it takes the light to get from the target body to the observer. This can be large for a target body such as Jupiter near conjunction with the Sun, so it's much better to do this from a barycentric perspective. That takes the Earth's orbital velocity out of the equation.

Reply:
For other than the time delay, that is.

Reply:
Thanks for the comments @mkbrewer - I must admit I'm getting quite confused by all the different possible time delays. I'm also confused by what you mean when you say a "topocentric perspective". All the vectors in ```erfa.ld``` are barycentric - it's just a question of at what time these positions correspond to.

My understanding is below. Assuming the light ray from the body reaches the observer (Earth) at time `t`:

- `p`: vector from observer to location of source *when it emitted ray* that reaches observer at `t`
- `e`: vector from Sun to observer *at the time of closest approach of the light ray to the Sun*
- `q`: vector from Sun to source *at the time when the source emitted ray* that reaches observer at `t`

In my code I'm ignoring the motion of the sun around the Barycentre and using the position at time `t`, but in the terms I use above, could you point out where I'm going wrong?

Reply:
Sorry, my comment concerning the Earth's orbital velocity isn't relevant. Both `eh` and `eb` are calculated without any time delay. I hadn't thought about this since May and was not fully back up to speed  when I said that. Regarding `e`, the time of closest approach is mainly for gravitational deflection by planets. Jupiter, for instance, can move 12 arcseconds wrt the solar system barycenter in the time that it takes for light to travel from it to Earth. The Sun moves much more slowly. In fact, it is stated in the Explanatory Supplement that just using `eh` without any time delay can only cause an error in the deflection of up to 0.1 mas which is less than the 0.5 mas accuracy of the formula itself. My main concern here is the lack of time delay in the position of the Sun used in `q`. Just scaling up that 0.1 mas from 8.3 minutes to the 4 hours that it takes light to travel from Neptune yields 2.9 mas.  What I meant by topocentric perspective is that in computing `q` you are using two vectors that involve the position of the Earth rather than the solar system barycenter. One of these, `eh` is geocentric, but the other `astrometric` (now `srepr`) can be topocentric which can displace the target body by up to 6378 km. For the Moon which has an orbital velocity of 1 km / s about the Earth, that translates into 106 minutes of time or about 53 arcminutes in position relative to the Earth. I don't know by how much these approximations affect the gravitational deflection, but you can find out yourself by implementing the algorithm that I suggested below. It is the algorithm that both the ES and I use, so comparing the results of that algorithm to the results from this PR should give you a good idea of the error.